File: singletonShopCart.py
class Carrinho:

    __instance = None

    def __init__(self):
        if Carrinho.__instance is None:
            Carrinho.__instance = self
            self.items = []
        else:
            raise Exception("Esta classe é um singletonn")
        
    @staticmethod
    def get_instance():
        if not Carrinho.__instance:
            Carrinho()
        
        return Carrinho.__instance
    
    def add_item(self, item):
        self.items.append(item)

    def get_item(self):
        return self.items


class Produto:

    def __init__(self, nome, preco):
        self.nome = nome
        self.preco = preco


#cliente:

#Utilizando o Singleton para criar o carrinho:
carrinho1 = Carrinho.get_instance()

#Criar os produtos:
produto1 = Produto("Banana", 4.00)
produto2 = Produto("Maçã", 7.99)
produto3 = Produto("Melão", 5.89)
produto4 = Produto("Repolho", 3.79)
produto5 = Produto("Água", 3.79)
produto6 = Produto("Linguiça", 3.79)

#Adicionar itens no carrinho:
carrinho1.add_item(produto1)
carrinho1.add_item(produto2)
carrinho1.add_item(produto3)
carrinho1.add_item(produto4)
carrinho1.add_item(produto5)
carrinho1.add_item(produto6)

#Exibindo a lista total dos carrinhos:
print("----------- Carrinho 1 -----------")
for item in carrinho1.get_item():
    print(item.nome, " - ", item.preco)
    numItens = numItens+1
    total = total + item.preco

File: singletonShopCart.js
const Carrinho = (function(){
    let instance;

    function createInstance(){
        let produtos = [];

        function addProduto(produto){
            produtos.push(produto);
        }

        function getProduto(){
            return produtos;
        }

        function clearCarrinho(){
            produtos = [];
        }

        return{
            addProduto,
            getProduto,
            clearCarrinho

        };
    }

    return{
        getInstance: function(){
            if(!instance){
                instance = createInstance();
            }
            return instance;
        }
    }
})()

const carrinho = Carrinho.getInstance();

carrinho.addProduto({id: 001, nome: "Produto 1", preco: 10});
carrinho.addProduto({id: 002, nome: "Produto 2", preco: 20});
carrinho.addProduto({id: 003, nome: "Produto 3", preco: 30});
carrinho.addProduto({id: 004, nome: "Produto 4", preco: 40});
carrinho.addProduto({id: 005, nome: "Produto 5", preco: 50});
carrinho.addProduto({id: 006, nome: "Produto 6", preco: 60});
carrinho.addProduto({id: 007, nome: "Produto 7", preco: 70});

console.log(carrinho.getProduto());
carrinho.clearCarrinho();
console.log(carrinho.getProduto());